### PR TITLE
[DROOLS-6432] Calling fireAllRules simultaneously from multiple threa…

### PR DESCRIPTION
**Pull request by tkobayas**
…ds sometimes causes an exception inside MVELConsequence#evaluate.

### CODE REVIEW

**Comment by tkobayas** [comment]
Firstly I used Semaphore or CountDownLatch but getValue() could be called recursively so I ended up with ReentrantLock.

File: src/test/java/org/mvel2/tests/perftests/ConcurrentEvaluationWithContextTests.java [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+
+            parserContext.addIndexedInput(names);
+            parserContext.addInput("result", List.class);
+            parserContext.addInput("$bus", Bus.class);
+            parserContext.addInput("$title", String.class);
+
+            expression = MVEL.compileExpression("result.add($bus.karaoke.dvd[$title].artist);", parserContext);
+
+        } catch (Exception e) {
+            e.printStackTrace();
```
Comment (MalcolmOdd) [inline src/test/java/org/mvel2/tests/perftests/ConcurrentEvaluationWithContextTests.java:59]: You might as well let it through by adding a `throws` clause. The test won't succeed if it gets here.

File: src/test/java/org/mvel2/tests/perftests/ConcurrentEvaluationWithContextTests.java [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+
+    @Test(timeout = 10000)
+    public void testDynamic() throws Exception {
+        OptimizerFactory.setDefaultOptimizer(OptimizerFactory.DYNAMIC);
+        internalConcurrentEvaluation();
+    }
+
+    @Test(timeout = 10000)
+    public void testASM() throws Exception {
+        OptimizerFactory.setDefaultOptimizer("ASM");
```
Comment (MalcolmOdd) [inline src/test/java/org/mvel2/tests/perftests/ConcurrentEvaluationWithContextTests.java:83]: Awesome! Testing this kind of thing with ASM is super-important!

File: src/test/java/org/mvel2/tests/perftests/ConcurrentEvaluationWithContextTests.java (hunk 1 above)
Comment (tkobayas) [reply src/test/java/org/mvel2/tests/perftests/ConcurrentEvaluationWithContextTests.java]: Thank you for the advice! Fixed.

**Pull request closed without merging**